master/internal/provisioner/scale_decider.go

```diff
@@ -98,78 +100,44 @@ func (s *scaleDecider) updateInstanceSnapshot(instances []*Instance) bool {
 func (s *scaleDecider) findInstancesToTerminate(
 	maxInstanceNum int,
 ) []string {
-	idleAgents := s.schedulerSnapshot.Agents
-	instances := s.instanceSnapshot
+	toTerminate := make(map[string]bool)
+	idleInstances := make(map[string]bool)
 
-	// Terminate stopped instances.
-	stoppedInstanceIDs := make([]string, 0, len(instances))
-	candidates := make([]*Instance, 0, len(instances))
-	for _, inst := range instances {
+	// Terminate stopped instances and find idle instances.
+	for _, inst := range s.instanceSnapshot {
 		switch inst.State {
 		case Stopped:
-			stoppedInstanceIDs = append(stoppedInstanceIDs, inst.ID)
-		case Starting, Running:
-			candidates = append(candidates, inst)
-		}
-	}
-	instances = candidates
+			toTerminate[inst.ID] = true
 
-	// We assume that there are no duplicate instance IDs in the input.
-	// Separate out unique agents. Duplicate agents are not handled.
-	uniqueIdleAgents := make(map[string]*scheduler.AgentSummary)
-	for _, agent := range idleAgents {
-		if _, ok := uniqueIdleAgents[agent.Name]; !ok {
-			uniqueIdleAgents[agent.Name] = agent
-		}
-	}
-
-	// Find instances to terminate.
-	toMark := make(map[string]bool)
-	uniqueAgentNames := make(map[string]*Instance)
-	for _, inst := range instances {
-		switch first, ok := uniqueAgentNames[inst.AgentName]; {
-		case ok:
-			toMark[inst.ID] = true
-			toMark[first.ID] = true
-		case inst.AgentName == "":
-			toMark[inst.ID] = true
-		default:
-			uniqueAgentNames[inst.AgentName] = inst
-		}
-	}
-	for _, inst := range instances {
-		if _, ok := uniqueIdleAgents[inst.AgentName]; ok {
-			toMark[inst.ID] = true
+		case Running:
+			if _, ok := s.agentSnapshot[inst.AgentName]; ok {
+				idleInstances[inst.ID] = true
+			}
 		}
 	}
 
-	// Mark instances to terminate for some time before actually terminating them.
-	toTerminate := make(map[string]bool)
-	instancesMarked := make(map[string]time.Time)
-	now := time.Now()
-	for instID := range toMark {
-		switch t, ok := s.instancesMarked[instID]; {
-		case ok && now.After(t.Add(s.maxIdleAgentPeriod)):
-			toTerminate[instID] = true
-		case ok:
-			instancesMarked[instID] = t
-		default:
-			instancesMarked[instID] = now
-		}
+	// Terminate instances that are idle for a long time.
+	var longLived map[string]bool
+	s.idleInstances, longLived = findLongLived(s.idleInstances, idleInstances, s.maxIdlePeriod)
```

Comment:
?: `longLived`-> `longIdle` (don't love this name either but is more accurate imo).